### ISSUE TITLE
investigate 2.13 community build breakage for 2.13.0-RC1

### ISSUE THREAD

**Issue by SethTisue** [scala/bug#11453]
current nightly: ~~2.13.0-pre-59975bb~~ ~~2.13.0-pre-af24410~~ 2.13.0-pre-b4926ef

I did a sweep through the latest run log at https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/ , findings as follows (numbers are "number of downstream projects blocked")

advice for those who want to help is now up at https://github.com/scala/community-builds/wiki/Troubleshooting-a-failure . I can expand and update it in response to questions. you can also help simply by eyeballing a failure and investigating it by reading code, and posting your findings here

_might possibly_ have hit as-yet-unknown 2.13 regressions:

* [ ]  3  "http4s-parboiled2",  // diverging implicit expansion -- could be a Scala regression, looks tricky to troubleshoot
* [ ]   1 "circe-jackson",  // appears to have run afoul of recent (Feb? Mar?) changes to overloading resolution? could possibly be a Scala regression
* [ ]   1 "parboiled2",  // same as http4s-parboiled2

I also found some projects that are only failing for trivial-looking reasons. we could submit upstream fixes, or just put fixes or workarounds in our forks, and see if that uncovers any real issues (or, in some cases, unblocks downstream stuff):

* [ ]  10  "monix"  // compile errors because of new `PartialFunction` overload, should be easy workaround
* [ ] 3   "silencer",  // test failure, maybe just a changed error message wording?
* [ ]  3  "scala-logging",  // runs afoul of new rules about varargs overloading
* [ ] 2   "grizzled",  // M5 compat is in place, but there are collections-related compile errors, maybe just needs post-M5 tweaks?
* [ ] 0   "scalajson",  // looks like it might be hitting Som's change where `import` shadows local identifiers?
* [ ]   0 "scallop",  // trivial post-M5 `remove` vs `removed` fix needed

and then this is a dilemma:

* [ ]  13  "scalapb",  // dilemma: on master they have M5 support, but they also require Fastparse 2. what to do?

fixed

* [x]  9  "scopt",  // no arguments allowed for nullary method   thank you @xuwei-k
* [x]  2 "specs2-more",  // overly sensitive test  thank you @etorreborre 
* [x]  1  "base64",  // needs workaround for https://github.com/scala/bug/issues/11158 which won't be fixed in RC1   thank you @xuwei-k and @marklister
* [x]  1  "scalariform",  // looks like a trivial compile error perhaps related to removal of `foo bar ()` syntax  thank you @Philippus and @godenji 
* [x] 0 scala-parallel-collections thank you @eed3si9n 
* [x] 0 scala-collection-laws  scala/community-builds@53ff214548d20b0d09837fef577115ec083a2c70
* [x] 0   "twotails",  // "no arguments allowed for nullary method"  thank you @Philippus 
* [x]   0 "scodec",  // we just needed to use the branch where 2.13 support is present
* [x]  0  "airframe",  // serialization related test failures, thank you @xerial 

not green, but investigated, and no unreported Scala regression seems present

* [ ] 0 scala-js    scala/community-builds@e5c65e2f38bd8b0e7eb50af6282cb119981214a5  thank you @adriaanm
* [ ]   0 "kittens",  // hashing-related test failure  thank you @joroKr21 
* [ ]  0  "scala-swing",  // looks like it needs tweaks for post-M5 changes  thank you @Sciss

**Comment by joroKr21** [comment]
I would like to check the `kittens` issue if no one beats me to it. How can I reproduce it?

**Comment by SethTisue** [comment]
I’m writing up some advice on this and I’ll link it here in a little while

**Comment by SethTisue** [comment]
okay, advice for those who want to help is now up at https://github.com/scala/community-builds/wiki/Troubleshooting-a-failure

you can also help simply by eyeballing a failure and investigating it by reading code, and posting your findings here

**Comment by SethTisue** [comment]
@etorreborre what do you make of the specs2-more failure? 

```
[specs2-more] [error]   x h2 id="1"/> ++ <h3/>).updateHeadAttribute("id", 3) === (<h2 id='3'/> ++ <h3/>)
[specs2-more] [error]  <h2 id="3"/><h3/> != List(<h2 id="3"/>, <h3/>) (HtmlxSpec.scala:15)
[specs2-more] [info]   + h2>hello</h2>.addHeadersAnchors.toString must beMatching("<a name=\"hello\"><h2>hello</h2></a>")
[specs2-more] [info]   the headers methods
[specs2-more] [error]     x collects all headers of a document
[specs2-more] [error]  <h1>title1</h1><h2>title2</h2> != List(<h1>title1</h1>, <h2>title2</h2>) (HtmlxSpec.scala:50)
```

**Comment by adriaanm** [comment]
I fixed the scala-js issue in https://github.com/scalacommunitybuild/scala-js/tree/community-build-2.13 but forgot to update the ref I guess

**Comment by SethTisue** [comment]
@adriaanm I already pushed scala/community-builds@e5c65e2f38bd8b0e7eb50af6282cb119981214a5

**Comment by Jasper-M** [comment]
> base64", // StringBuilder stopped allowing assignment to its .length member? is that an intentional change we made?

This is #11158

**Comment by SethTisue** [comment]
> This is #11158

@adriaanm says a proper fix isn't possible right now, so we'd have to just work around it in base64. (but note that base64 isn't blocking any downstream projects.)

**Comment by etorreborre** [comment]
@SethTisue I tried a small change to the expectations in `HtmlxSpec`: b3649e9e7. Let's see how it goes

**Comment by SethTisue** [comment]
@etorreborre that does it, thanks!

```
[info] Project specs2----------------------: SUCCESS (project rebuilt ok)
[info] Project specs2-more-----------------: SUCCESS (project rebuilt ok)
```

**Comment by joroKr21** [comment]
The failing tests in `kittens` test that the derived `Hash` is the same as Scala's `Hashing` and `hashCode`. I guess the cause is scala/scala#7693.

But I'm not sure:
  *  Why do we need `Hash` derivation if the only allowed implementation is the same as `hashCode`?
  * How to fix it? I guess have a special instance when `A <: Product` but what about cross compilation?

**Comment by Sciss** [comment]
I don't really understand the __scala-swing errors__ in https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1895/consoleFull
They all seem to stem from the fact that sbt doesn't see `MapWrapper`, `SetWrapper`, `BufferWrapper`. But they are in the sources: https://github.com/scala/scala-swing/blob/work/src/main/scala-2.13.0-M5/scala/swing/MapWrapper.scala

Could the problem be that sbt is not assuming source directory `src/main/scala-2.13.0-M5` simply because __the version is no longer `-M5`__?

**Comment by smarter** [comment]
> Could the problem be that sbt is not assuming source directory src/main/scala-2.13.0-M5 simply because the version is no longer -M5?

Yup. To get something more fine-grained you can set up your own rules in your build, e.g.: https://github.com/akka/akka/blob/65ccada280cf4a08701b0b05f7efab46560447ff/project/AkkaBuild.scala#L128-L137

**Comment by Sciss** [comment]
Ok, we can adopt that. Should I just push another commit to scala-swing repo with this change?

**Comment by SethTisue** [comment]
> Should I just push another commit to scala-swing repo with this change?

Yes please.

(At least, unless it seems to you, having looked at the code that involved, that there's practically zero chance of there being any Scala regression involved that would require delaying RC1.)

**Comment by SethTisue** [comment]
I've bumped the community build SHA to 2.13.0-pre-59975bb, just in case the scala/bug#11450 fix being in might uncover something else downstream.

**Comment by Sciss** [comment]
@SethTisue done in https://github.com/scala/scala-swing/commit/732c1cc781f3199f3cc8efccc3cc72c85118208d

**Comment by SethTisue** [comment]
@Sciss thanks, now I get a different error

```
[scala-swing] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-swing-05f61027f8368c12004507d9dfeb2e29933c4d82/src/main/scala-2.13+/scala/swing/BufferWrapper.scala:51:16: method patchInPlace overrides nothing.
[scala-swing] [error] Note: the super classes of class BufferWrapper contain the following, non final members named patchInPlace:
[scala-swing] [error] def patchInPlace(from: Int, patch: scala.collection.IterableOnce[A], replaced: Int): scala.swing.BufferWrapper[A]
[scala-swing] [error]   override def patchInPlace(from: Int, patch: scala.collection.Seq[A], replaced: Int): this.type = {
[scala-swing] [error]                ^
[scala-swing] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-swing-05f61027f8368c12004507d9dfeb2e29933c4d82/src/main/scala/scala/swing/Container.scala:28:21: class Content needs to be abstract. Missing implementation for:
[scala-swing] [error]   def patchInPlace(from: Int, patch: scala.collection.IterableOnce[scala.swing.Component], replaced: Int): Content.this.type // inherited from trait Buffer 
[scala-swing] [error] (Note that scala.collection.IterableOnce[A] does not match scala.collection.Seq[A]: trait Seq in package collection is a subclass of trait IterableOnce in package collection, but method parameter types must match exactly.)
[scala-swing] [error]     protected class Content extends BufferWrapper[Component] {
[scala-swing] [error]                     ^
[scala-swing] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-swing-05f61027f8368c12004507d9dfeb2e29933c4d82/src/main/scala/scala/swing/TabbedPane.scala:82:10: object creation impossible. Missing implementation for:
[scala-swing] [error]   def patchInPlace(from: Int, patch: scala.collection.IterableOnce[scala.swing.TabbedPane.Page], replaced: Int): scala.swing.TabbedPane.pages.type // inherited from trait Buffer 
[scala-swing] [error] (Note that scala.collection.IterableOnce[A] does not match scala.collection.Seq[A]: trait Seq in package collection is a subclass of trait IterableOnce in package collection, but method parameter types must match exactly.)
[scala-swing] [error]   object pages extends BufferWrapper[Page] {
[scala-swing] [error]          ^
```

looks like a normal consequence of https://github.com/scala/scala/pull/7155 / https://github.com/scala/scala/pull/7158  ?

**Comment by SethTisue** [comment]
@xerial wdyt about the airframe failure, could it be a regression in Scala?

```
[airframe] [info] JDBCCodecTest:
[airframe] [info] - ResultSet to JSON maps *** FAILED ***
[airframe] [info]   "{"id":[2,"name":"yui]"}" was not equal to "{"id":[1,"name":"leo]"}" (JDBCCodecTest.scala:187)
```

**Comment by Sciss** [comment]
@SethTisue yes I think so; we can patch this when the RC1 artifact is available, I guess?

**Comment by SethTisue** [comment]
@Sciss yup

**Comment by xerial** [comment]
Let me check about airframe failure

**Comment by xerial** [comment]
@SethTisue Yeah. This is possibly a regression around TraversableOnce.toIndexedSeq or Iterator in Scala collection.

**Comment by xerial** [comment]
It seems Iterator's hasNext() is called twice in some code, so if implementation like JDBC's ResultSet.next() that changes the internal state, some elements in an Iterator is skipped at toIndexedSeq. I'll file this as another ticket.

**Comment by xerial** [comment]
@SethTisue Reproduced this bug and filed at https://github.com/scala/bug/issues/11455

**Comment by eed3si9n** [comment]
scala-parallel-collections: Thanks to https://github.com/scala/scala/pull/7940, scala-parallel-collections seems to work fine locally using 2.13.0-pre-59975bb.

**Comment by eed3si9n** [comment]
scopt: @xuwei-k sent me a PR, which should fix it - https://github.com/scopt/scopt/pull/240

**Comment by Philippus** [comment]
base64 seems to already have a workaround by @xuwei-k: https://github.com/marklister/base64/pull/19

**Comment by xerial** [comment]
@SethTisue Fixed this issue at airframe-side for now https://github.com/wvlet/airframe/commit/6c542eb36188293491f9754cc9b010d95bbdff29

**Comment by Philippus** [comment]
scala-xml-quote fix: https://github.com/densh/scala-xml-quote/pull/18, seems the community build uses the fork at https://github.com/allanrenucci/scala-xml-quote though.

**Comment by Philippus** [comment]
I think this should fix the scalariform error: https://github.com/scala-ide/scalariform/pull/276

**Comment by SethTisue** [comment]
@joroKr21 the kittens thing looks like it might better be discussed on a kittens ticket. I've moved kittens to the "not green, but investigated, and no unreported Scala regression seems present" section

**Comment by SethTisue** [comment]
in airframe, I'm now getting serialization related test failures:

```
[airframe] [error] Failed tests:
[airframe] [error] 	wvlet.airframe.SerializationTest
[airframe] [error] 	wvlet.airframe.DesignSerializationTest
[airframe] [error] 	wvlet.airframe.ProviderSerializationTest
```

this was in a local test but the same error should show up in https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1902/

**Comment by Philippus** [comment]
twotails fix: https://github.com/wheaties/TwoTails/pull/34

**Comment by SethTisue** [comment]
thanks everyone for all the excellent help! as they come in, I'm pulling in fixes, marking comments as "resolved", and updating the statuses above

**Comment by SethTisue** [comment]
@Philippus next scala-xml-quote errors are

```
[scala-xml-quote] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-xml-quote-2ba1cf1077e723d34a4d21d5a898d1ac25fc1aeb/src/test/scala/scala/xml/quote/SimpleNodeSuite.scala:6:30: value ≈ is not a member of scala.xml.NodeSeq
[scala-xml-quote] [error]     assert(xml"<foo/><bar/>" ≈ <foo/><bar/>)
[scala-xml-quote] [error]                              ^
[scala-xml-quote] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-xml-quote-2ba1cf1077e723d34a4d21d5a898d1ac25fc1aeb/src/test/scala/scala/xml/quote/SimpleNodeSuite.scala:109:66: value ≈ is not a member of scala.xml.NodeSeq
[scala-xml-quote] [error]     assert(xml"<![CDATA[hello, world]]><![CDATA[hello, world]]>" ≈
[scala-xml-quote] [error]                                                                  ^
[scala-xml-quote] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scala-xml-quote-2ba1cf1077e723d34a4d21d5a898d1ac25fc1aeb/src/test/scala/scala/xml/quote/UnquoteSuite.scala:20:23: type mismatch;
[scala-xml-quote] [error]  found   : scala.xml.NodeBuffer
[scala-xml-quote] [error]  required: String | Seq[scala.xml.Node] | Option[Seq[scala.xml.Node]]
[scala-xml-quote] [error]        
[scala-xml-quote] [error]     assert(xml"<foo a=${<bar/><bat/>}/>" ≈ <foo a={<bar/><bat/>}/>)
[scala-xml-quote] [error]                       ^
```

**Comment by godenji** [comment]
> I think this should fix the scalariform error: [scala-ide/scalariform#276](https://github.com/scala-ide/scalariform/pull/276)

Scalariform 0.2.8 released, should be available in 30 minutes or so...

**Comment by Philippus** [comment]
scapegoat: https://github.com/scalacommunitybuild/scapegoat/pull/1, hopefully it just works without this option enabled.

**Comment by xerial** [comment]
@SethTisue That serialization error seems related to https://github.com/scala/scala/pull/7624.

**Comment by SethTisue** [comment]
@Philippus it appears from the next error that scapegoat hasn't moved to the new collections yet:

```
[scapegoat] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scapegoat-faa3d8e37f7954c9d1aabc690b81dc42fa649c70/src/main/scala/com/sksamuel/scapegoat/Feedback.scala:17:61: type mismatch;
[scapegoat] [error]  found   : scala.collection.mutable.ListBuffer[com.sksamuel.scapegoat.Warning]
[scapegoat] [error]  required: Seq[com.sksamuel.scapegoat.Warning]
[scapegoat] [error]   def warnings(level: Level): Seq[Warning] = warnings.filter(_.level == level)
[scapegoat] [error]                                                             ^
[scapegoat] [error] /Users/tisue/community.213/target-0.9.16/project-builds/scapegoat-faa3d8e37f7954c9d1aabc690b81dc42fa649c70/src/main/scala/com/sksamuel/scapegoat/Feedback.scala:17:61: type mismatch;
[scapegoat] [error]  found   : scala.collection.mutable.ListBuffer[com.sksamuel.scapegoat.Warning]
[scapegoat] [error]  required: Seq[com.sksamuel.scapegoat.Warning]
[scapegoat] [error]   def warnings(level: Level): Seq[Warning] = warnings.filter(_.level == level)
[scapegoat] [error]                                                             ^
```

**Comment by SethTisue** [comment]
for scodec, we are using the series/1.10.x branch, but the series/1.11.x branch is where 2.13 support is. so I think the next step would be to try that branch instead, and then also consider back porting the change to  the 2.12 community build

**Comment by xerial** [comment]
@SethTisue Fixed the serialization error of airframe by forking the JVM: https://github.com/wvlet/airframe/commit/4f9dc062310133ea76b7013951d6b669e01ed9ae

The cause is sbt's class loader issue, which doesn't work well with Java's ObjectSerializer to serialize Scala 2.13.0's new collections. This issue is already addressed in sbt-1.3.0's layered class loader (cc: @eed3si9n)

**Comment by SethTisue** [comment]
@xerial yup, that does it

94 green projects now :+1:

**Comment by SethTisue** [comment]
removed the "blocker" label, as we're probably in good enough shape for the RC1 release now, though work here will continue up through the 2.13.0 release and beyond

but there remains the blocker issue of pushing scala-java8-compat removal through the community build; tracking that separately at https://github.com/scala/bug/issues/11452

**Comment by Philippus** [comment]
looks like these need quite some work to go to scala 2.13.x:
"scala-xml-quote" (would need to go to fastparse 2 I think), "scapegoat" and "scribe"

**Comment by SethTisue** [comment]
@Philippus thanks for checking on it. I removed them from the list.

**Comment by Philippus** [comment]
@SethTisue I was looking at "scala-logging", I think the specs could be fixed with some clever annotations or `asInstanceOf`-stuff.
But I think it requires a nice proper fix, as it will probably hit real end-users of "scala-logging" eventually.
Maybe open an issue on the project?

**Comment by SethTisue** [comment]
@Philippus yes please :-)

(I would eventually get to it myself, but not this week)

**Comment by SethTisue** [comment]
>   3 "http4s-parboiled2", // diverging implicit expansion -- could be a Scala regression, looks tricky to troubleshoot
>  1 "parboiled2", // same as http4s-parboiled2 

this is not eyeball-able, too hairy to troubleshoot ourselves, let's not hold the release up over it

> 1 "circe-jackson", // appears to have run afoul of recent (Feb? Mar?) changes to overloading resolution? could possibly be a Scala regression

also hairy, very likely fixable in a minor version even if turns out to be Scala's fault, ditto and not holding things up

**Comment by SethTisue** [comment]
there have been 3 regressions:

* [x] unexpected FAILED: scala-parallel-collections  — fixed, thanks @adriaanm 
* [ ] unexpected FAILED: jackson-module-scala
* [ ] unexpected FAILED: spire

Adriaan and/or the rest of the team are looking into the remaining two

**Comment by sjrd** [comment]
scala-js should be back on track, now. Let me know if something breaks again.

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1923/ running on 2.13.0-pre-63769af is `SUCCESS 94 UNEXPECTED 0`, so we're good to go again for RC1, knock on wood

@sjrd will do (after I unfreeze all the project SHAs after RC1 is out)

**Comment by SethTisue** [comment]
Adriaan suggested testing scala-js now rather than waiting, so I pushed https://github.com/scala/community-builds/commit/3306bd8de569b005a937777715a38de94c56d18a and to get fast feedback did a local, scala-js-only run. there were 2 test failures: https://gist.github.com/SethTisue/d83b051d0be8890346ad8612122aa245

the tests are testing error messages, and it looks plausible to me that the additional errors the compiler are now generating are probably reasonable errors coming from @adriaanm's recent rechecks changes; or if they do indicate that refchecks is now a bit overzealous in a few cases, it's probably not a release blocker?

**Comment by sjrd** [comment]
Oh I had forgotten that the community build performed the `compiler/test`. Yeah, those were broken ~2 weeks ago, and I haven't had time to investigate them. I don't consider them a blocker, but I can have a look today.

**Comment by adriaanm** [comment]
Yeah, the error messages (I assume) are related to refchecks, which now checks all pairs in the override chain.

**Comment by sjrd** [comment]
OK, I can now confirm that the new error messages are progressions.

**Comment by SethTisue** [comment]
* reported circe-jackson failure: https://github.com/circe/circe-jackson/issues/54
* ticket on http4s failure: https://github.com/sirthias/parboiled2/pull/200

**Comment by SethTisue** [comment]
I updated https://github.com/scala/community-builds/blob/2.13.x/report/Report.scala a day or two ago to include all the failure reasons (to the extent they're known)
